### ISSUE TITLE
[Feature] [Training] FP8/MXFP8 grouped-GEMM API for MoE training

### ISSUE THREAD

**Issue by jayfurmanek** [ROCm/aiter#3038]
## Summary

  We're integrating AITER as the low-precision grouped-GEMM backend for a "lite" build of [TransformerEngine](https://github.com/ROCm/TransformerEngine) that drops the C++ extension. BF16/FP16 works today via `aiter.gmm` / `aiter.ptgmm`. **FP8  and MXFP8 are blocked**: AITER's existing 8-bit MoE kernels (`moe_gemm_a8w8`, moe_gemm_a8w8_blockscale`, and the MX siblings) are inference-shaped — forward-only, router-driven, internally allocated outputs, scale layouts not aligned with TE's recipes.

  This issue proposes a stable, training-ready entry point: `aiter.grouped_gemm_fp8`, covering both per-tensor / per-row / 1×128-block FP8 and 1×32-block MXFP8 under a single dispatcher.

  ## Context

  Each TE `GroupedLinear` (used by Megatron `GroupedMLP`) issues **three** grouped-GEMM calls per training step — one forward, one dgrad, one wgrad:

  | Pass  | Layout | A shape (weights or inputs) | B shape     | Output shape   |
  | ----- | ------ | --------------------------- | ----------- | -------------- |
  | fwd   | `TN`   | weights `[E, N, K]`         | `[ΣM, K]`   | `[ΣM, N]`      |
  | dgrad | `NN`   | weights `[E, N, K]`         | `[ΣM, N]`   | `[ΣM, K]`      |
  | wgrad | `NT`   | inputs `[ΣM, K]`            | `[ΣM, N]`   | `[E, N, K]`    |

  Per-expert bias is added in fwd; dgrad runs with `single_output=True` (one packed output, not a per-expert list); wgrad runs with `accumulate=True` to fuse into Megatron's `main_grad`.

  `A` and `B` arrive as `List[Float8Tensor]` or `List[MXFP8Tensor]` (one per expert), each carrying its own scale. Outputs are caller-preallocated. There is no router — just `m_splits: List[int]`.

  ## What's missing

  To support FP8 / MXFP8 MoE training, AITER needs (in roughly priority order):

  1. **Backward GEMMs** — `NN` (dgrad) and `NT` (wgrad) variants of the 8-bit grouped kernel. Today only `TN` forward exists, so 8-bit training is impossible.
  2. **Caller-provided `out` + `accumulate=True`** — wgrad fuses into Megatron's `main_grad`. The kernel must do `out += A @ B` into an externally allocated tensor; today `allocate_output` always returns a fresh buffer.
  3. **Routing-free invocation** — accept `m_splits: List[int]` (or precomputed int32 offsets) instead of requiring a `RoutingData.expt_data` produced by a router. `gather_indx = scatter_indx = None`, `n_expts_act = 1`, no `gammas`.
  4. **All five TE scale layouts** as a documented, dispatchable contract:

     | TE recipe                            | Scale dtype        | A scale shape          | B scale shape                  |
     | ------------------------------------ | ------------------ | ---------------------- | ------------------------------ |
     | DelayedScaling / per-tensor current  | `fp32`             | `[E]`                  | `[E]`                          |
     | Per-row current                      | `fp32`             | `[ΣM, 1]`              | `[E, N, 1]`                    |
     | `Float8BlockwiseQ` (1×128)           | `fp32`             | `[ΣM, ⌈K/128⌉]`        | `[E, ⌈K/128⌉, ⌈N/128⌉]`        |
     | **MXFP8 (1×32)**                     | **`uint8` (E8M0)** | **`[ΣM, ⌈K/32⌉]`**     | **`[E, ⌈K/32⌉, ⌈N/32⌉]`**      |
     | MXFP8 (1×32), CDNA4 swizzled         | `uint8` (E8M0)     | swizzled (TE layout)   | swizzled (TE layout)           |

     Today `_moe_gemm_a8w8` accepts only **scalar** static scales, not per-expert `[E]`. The MX variants exist for forward but the swizzle helper (`unswizzle_mx_scale_cdna4`, `_triton_kernels/moe/moe_op_gemm_a8w8.py:95`) is forward-only — for
  wgrad/dgrad the swizzle has to round-trip cleanly when the scale tensor is read along the contraction dim instead of the output dim.

  5. **Empty-group safety** — `sum(m_splits) == 0` and per-expert `hist[e] == 0` must be no-ops, not asserts. Megatron legitimately routes 0 tokens to local experts before load-balancing converges.
  6. **Stable, versioned export** from `aiter.__init__`. The current `moe_gemm_a8w8` / `_blockscale` / MX kernel signatures are kernel-shaped and not part of the public API.

  ## Proposed API

  ```python
  # Public surface: aiter.grouped_gemm_fp8
  # Wrapper:        aiter/ops/triton/gemm/grouped/grouped_gemm_fp8.py
  # Kernels:        aiter/ops/triton/_triton_kernels/gemm/grouped/grouped_gemm_fp8.py

  from enum import Enum
  from typing import List, Optional
  import torch


  class FP8ScaleKind(Enum):
      PER_TENSOR  = "per_tensor"   # fp32 [E] per expert
      PER_ROW     = "per_row"      # fp32 [ΣM, 1] / [E, N, 1]
      BLOCK_1x128 = "block_1x128"  # fp32, Float8BlockwiseQ
      MX_1x32     = "mx_1x32"      # uint8 E8M0, MXFP8 (optionally CDNA4-swizzled)


  def grouped_gemm_fp8(
      a: torch.Tensor,                    # FP8 (E4M3/E5M2), shape per layout
      b: torch.Tensor,                    # FP8, shape per layout
      out: torch.Tensor,                  # PRE-ALLOCATED, shape per layout
      m_splits: torch.Tensor,             # int32 [E], sum(m_splits) == ΣM
      *,
      scale_kind: FP8ScaleKind,
      a_scale: torch.Tensor,              # dtype/shape per scale_kind table
      b_scale: torch.Tensor,              # dtype/shape per scale_kind table
      mx_swizzle: bool = False,           # MX_1x32 only: scales arrive in CDNA4 swizzle
      layout: str = "TN",                 # "TN" (fwd), "NN" (dgrad), "NT" (wgrad)
      accumulate: bool = False,           # if True: out += A @ B
      use_split_accumulator: bool = True, # fp32 accumulation (TE recipe knob, unused for Triton)
      bias: Optional[torch.Tensor] = None,        # [E, N], fwd only
      out_dtype: torch.dtype = torch.bfloat16,    # bf16 / fp16 / fp32
      deterministic: bool = True,        # forbid split_k > 1 reductions
  ) -> None:
      """In-place grouped FP8 / MXFP8 GEMM for MoE training.

      For each expert e (with off(e) = cumulative m_splits prefix sum):
          layout="TN":  out[off(e):off(e+1), :] = b[off(e):off(e+1), :] @ a[e].T   (fwd)
          layout="NN":  out[off(e):off(e+1), :] = b[off(e):off(e+1), :] @ a[e]    (dgrad)
          layout="NT":  out[e]                  = a[off(e):off(e+1), :].T
                                                @ b[off(e):off(e+1), :]            (wgrad)

      accumulate=True turns assignment into +=. Empty groups are no-ops.
      Element dtype is FP8 (E4M3 / E5M2) for all scale_kinds; MX_1x32 differs
      only in the scale tensor dtype/layout (E8M0 uint8, optionally swizzled).
      No routing: gather/scatter and top-k weights are the caller's job.
      """
  ```

  ### Output shapes

  | layout | `a`         | `b`         | `out`       |
  | ------ | ----------- | ----------- | ----------- |
  | `TN`   | `[E, N, K]` | `[ΣM, K]`   | `[ΣM, N]`   |
  | `NN`   | `[E, N, K]` | `[ΣM, N]`   | `[ΣM, K]`   |
  | `NT`   | `[ΣM, K]`   | `[ΣM, N]`   | `[E, N, K]` |

  ### Helpers

  ```python
  # Pack per-expert scales from a list of Float8Tensor / MXFP8Tensor
  # (zero-copy when contiguous in expert order).
  def pack_per_expert_scales(scales: List[torch.Tensor]) -> torch.Tensor: ...

  # Convert host-side List[int] to the int32 offsets tensor the kernel consumes.
  def m_splits_to_offsets(m_splits: List[int], device: torch.device) -> torch.Tensor: ...

  # Apply / undo CDNA4 swizzle on an MX scale tensor.
  def mx_swizzle_scales(scales: torch.Tensor, *, undo: bool = False) -> torch.Tensor: ...
  ```

  ## Acceptance criteria

  - [ ] `layout="TN"`, `"NN"`, `"NT"` all implemented for FP8 *and* MXFP8 inputs.
  - [ ] `accumulate=True` bit-equivalent to `out_pre + A @ B`.
  - [ ] `out` always caller-provided; kernel never reallocates.
  - [ ] All five scale layouts (per-tensor, per-row, 1×128, MX 1×32 plain, MX 1×32 swizzled) match TE reference numerics within recipe tolerance.
  - [ ] MX 1×32 swizzle round-trips correctly through `NT` wgrad — i.e. the contraction-dim scale read is consistent with the output-dim scale write in `TN` fwd.
  - [ ] `sum(m_splits) == 0` and per-expert `m_splits[e] == 0` are no-ops.
  - [ ] `use_split_accumulator=True` accumulates in fp32.
  - [ ] `deterministic=True` forces stable reduction order.
  - [ ] Exported from `aiter.__init__` with a versioned signature.

  ## Why not extend `fused_moe`, `moe_gemm_a8w8`, or `moe_op_mxfp4` directly

  `fused_moe` bundles routing + activation + 2-stage MLP — TE separates `fc1` / `fc2` into two `GroupedLinear` modules with independent gradient paths. `moe_gemm_a8w8` and the MX kernels are closer but forward-only and keyed off `RoutingData`;
  wiring TE-lite directly into them would force every consumer to duplicate `_compute_expt_data_internal`, scale packing, and MX swizzle handling.

  A dedicated `grouped_gemm_fp8` leaves the inference-shaped MoE kernels untouched and gives any training framework — TE and others — a clean target across both FP8 and MXFP8.

  ## Environment

  - AITER: `<git-sha>`
  - TransformerEngine (lite branch): https://github.com/ROCm/TransformerEngine — `furmanek/dev-lite`
  - Megatron-LM with FP8 / MXFP8 MoE training, MI300X / MI325X / MI350
